### PR TITLE
export-config-list

### PR DESCRIPTION
**Pull request by who7708**
1. add server/node name. not support Trojan URL
2. export config list to file and save in document directory.

image

image

### CODE REVIEW

File: app/src/main/java/io/github/trojan_gfw/igniter/TrojanConfig.java [hunk 1]
```diff
@@ -79,6 +81,7 @@ public String generateTrojanConfigJSON() {
             return new JSONObject()
                     .put("local_addr", this.localAddr)
                     .put("local_port", this.localPort)
+                    .put("remote_server_name", this.remoteServerName)
```
Comment (wongsyrone) [inline app/src/main/java/io/github/trojan_gfw/igniter/TrojanConfig.java:84]: This doesn't belong to Trojan config.

File: app/src/main/java/io/github/trojan_gfw/igniter/servers/presenter/ServerListPresenter.java [hunk 2]
```diff
@@ -43,6 +46,52 @@ public void displayImportFileDescription() {
...
+        int index = 0;
+        for (TrojanConfig trojanConfig : trojanConfigs) {
+            JSONObject jsonObject = new JSONObject();
+            jsonObject.put("remarks", trojanConfig.getRemoteServerName());
+            jsonObject.put("server", trojanConfig.getRemoteAddr());
+            jsonObject.put("server_port", trojanConfig.getRemotePort());
+            jsonObject.put("password", trojanConfig.getPassword());
+            jsonObject.put("verify", trojanConfig.getVerifyCert());
+            // for future
+            // jsonObject.put("type", "trojan");
```
Comment (wongsyrone) [inline app/src/main/java/io/github/trojan_gfw/igniter/servers/presenter/ServerListPresenter.java:85]: We only support trojan, no other types.

File: app/src/main/java/io/github/trojan_gfw/igniter/TrojanConfig.java (hunk 1 above)
Comment (who7708) [reply app/src/main/java/io/github/trojan_gfw/igniter/TrojanConfig.java:84]: This field only mark the remote address in order to remember, which like the field in file, gui-config.json, download from the website.

Comment (who7708) [reply app/src/main/java/io/github/trojan_gfw/igniter/TrojanConfig.java:84]: > This doesn't belong to Trojan config.

This field only mark the remote address in order to remember, which like the field in file, gui-config.json, download from the website. What do you think?

File: app/src/main/java/io/github/trojan_gfw/igniter/servers/presenter/ServerListPresenter.java (hunk 2 above)
Comment (who7708) [reply app/src/main/java/io/github/trojan_gfw/igniter/servers/presenter/ServerListPresenter.java:85]: Ok，I‘ll delete the key named type

**Comment by TchaikovDriver** [comment]
LGTM.

File: app/src/main/java/io/github/trojan_gfw/igniter/TrojanConfig.java (hunk 1 above)
Comment (wongsyrone) [reply app/src/main/java/io/github/trojan_gfw/igniter/TrojanConfig.java:84]: How about renaming it to ServerRemark?

Comment (who7708) [reply app/src/main/java/io/github/trojan_gfw/igniter/TrojanConfig.java:84]: NP

File: app/src/main/java/io/github/trojan_gfw/igniter/servers/presenter/ServerListPresenter.java [hunk 3]
```diff
@@ -43,6 +46,51 @@ public void displayImportFileDescription() {
...
+                exportConfigFile();
+            }
+        });
+    }
+
+    private void exportConfigFile() {
+        boolean success = false;
+        try {
+            String exportContent = getExportContent();
+            success = TrojanHelper.writeStringToFile(exportContent, Globals.getIgniterExportPath());
```
Comment (TchaikovDriver) [inline app/src/main/java/io/github/trojan_gfw/igniter/servers/presenter/ServerListPresenter.java:63]: Noticed that you attempt to export file on external storage, which requires permission `WRITE_EXTERNAL_STORAGE`. You have to check whether the permission is granted using `ContextCompat.checkSelfPermission(Context,String)` and make permission request if necessary.
BTW, do not forget to declare `WRITE_EXTERNAL_STORAGE` permission in `AndroidManifest.xml`.

Comment (who7708) [reply app/src/main/java/io/github/trojan_gfw/igniter/servers/presenter/ServerListPresenter.java:63]: On Android 4.4 (API level 19) or higher, your app doesn't need to request any storage-related permissions to access app-specific directories within external storage. The `minimumSdkVersion` in build.gradle is 21, it'll work ok.

Comment (TchaikovDriver) [reply app/src/main/java/io/github/trojan_gfw/igniter/servers/presenter/ServerListPresenter.java:63]: > On Android 4.4 (API level 19) or higher, your app doesn't need to request any storage-related permissions to access app-specific directories within external storage. The `minimumSdkVersion` in build.gradle is 21, it'll work ok.

My bad, I forgot the api restriction, you are right.

**Comment by TchaikovDriver** [comment]
@wongsyrone LGTM.

**Pull request merged**